### PR TITLE
Fix [contrib:postgres_proxy] TLS issue when both upstream and downstream TLS are enabled

### LINKED ISSUE #27808: upstreamTLS for postgres_proxy works with clear text downstream, but not with tls secured client connection
setup : 

i am running 3 pods in kubernetes , 
	- client-pg-0 (running postgres 13.3, tls configured)
	- envoy-v25-0 (running envoy:contrib-v1.25-latest)
	- server-pg-tls-0 (running postgres 13.3, tls configured)

Issue : 

I want to connect to the server-pg from client-pg through envoy pod, and i want envoy to do the follwoing for me : 
- terminate tls that has been initiated from the client(downstream)
- establish tls secured connection with the server(upstream)

so i wrote down this following configuration and bootstraped my envoy pod with this following : 
```
  static_resources:
    listeners:
    - name: pg_listener
      address:
        socket_address:
          address: 0.0.0.0
          protocol: TCP
          port_value: 10000
      filter_chains:
      - filters:
        - name: envoy.filters.network.postgres_proxy
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.filters.network.postgres_proxy.v3alpha.PostgresProxy
            stat_prefix: egress_pg
            enable_sql_parsing: false
            terminate_ssl: true
            upstream_ssl: REQUIRE
        - name: envoy.filters.network.tcp_proxy
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
            stat_prefix: pg_tcp
            cluster: pg_cluster
        transport_socket:
          name: envoy.transport_sockets.starttls
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.transport_sockets.starttls.v3.StartTlsConfig
            tls_socket_config:
              common_tls_context:
                # tls_params:
                #   tls_minimum_protocol_version: TLSv1_2
                #   tls_maximum_protocol_version: TLSv1_2
                tls_certificates:
                - certificate_chain: {filename: "certs/tls.crt"}
                  private_key: {filename: "certs/tls.key"}
                validation_context:
                  trusted_ca:
                    filename: certs/ca.crt

    clusters:
    - name: pg_cluster
      type: STRICT_DNS
      load_assignment:
        cluster_name: pg_cluster
        endpoints:
        - lb_endpoints:
          - endpoint:
              address:
                socket_address:
                  address: server-pg-tls.default.svc
                  port_value: 5432
      transport_socket:
        name: envoy.transport_sockets.starttls
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.transport_sockets.starttls.v3.UpstreamStartTlsConfig
          tls_socket_config:
            allow_renegotiation: true
            common_tls_context:
              # tls_params:
              #   tls_minimum_protocol_version: TLSv1_2
              #   tls_maximum_protocol_version: TLSv1_2
              tls_certificates:
              - certificate_chain: {filename: "certs/tls.crt"}
                private_key: {filename: "certs/tls.key"}
              validation_context:
                trusted_ca:
                  filename: certs/ca.crt		
```

but, with this configuration I cannot achive my goal. it only works when i turn the ssl off with `sslmode=disable` from the client side. 

the following errors are shown when i try to create a tls secured connection from the client : 

client log : 
```
$ psql -d "user=postgres password=$pass dbname=postgres host=envoy-v25.default.svc port=10000 sslmode=require"
psql: error: SSL connection has been closed unexpectedly
```

envoy log : 
```
[2023-06-05 12:35:31.134][23][info][filter] [contrib/postgres_proxy/filters/network/source/postgres_filter.cc:255] [C2] postgres_proxy: upstream server rejected request to establish SSL connection. Terminating.
```

server log : 
```
2023-06-05 12:35:31.134 UTC [10169] LOG:  could not accept SSL connection: No error information
```

when i turn the ssl mode off from the client, it connects
```
$ psql -d "user=postgres password=$pass dbname=postgres host=envoy-v25.default.svc port=10000 sslmode=disable"
psql (13.2)
Type "help" for help.

postgres=# 

```
but as per doc api documentation suggests, this should work with `terminate_ssl:true`

more on that , if I only use the downstreamTLS and and no upstreamTLS and then connect to a tls insecured server , it works  fine. 

so , basically the downstreamTLS and upstreamTLS works fine with the postgre_proxy when used individually, but not simultineously for me.

please correct me if there is anything wrong with my configuration, or explain what is the issue here .

### PR DESCRIPTION
**Pull request by shiponcs**
Commit Message: Fix postgres_proxy tls issue when both upstream and downstream tls are enabled
Additional Description:
Risk Level: Low
Docs Changes: N/A
Release Notes: N/A
Fixes  #27808

### CODE REVIEW

**Comment by shiponcs** [comment]
CC: @cpakulski

**Comment by cpakulski** [comment]
I am away until beginning of September and will review this PR when I am
back.

File: contrib/postgres_proxy/filters/network/source/postgres_decoder.cc [hunk 1]
```diff
@@ -446,8 +446,18 @@ Decoder::Result DecoderImpl::onDataInNegotiating(Buffer::Instance& data, bool fr
...
+  /*
+   * (shiponcs):
+   * The length can be >= 1 for Write Buffer being not cleared,
+   * when upstream and downstream tls both are enabled,
+   * before receiving 'S' or 'N' from upstream for which
+   * the write buffer becomes "SN"/"SS" instead of "S"/"N".
+   * So we need to read the last(latest) byte from the buffer.
+   *
+   * */
+  if (data.length() >= 1) {
```
Comment (cpakulski) [inline contrib/postgres_proxy/filters/network/source/postgres_decoder.cc:459]: Can you verify that when length is 2 bytes, the content of it is SN or SS?

**Review by cpakulski — CHANGES_REQUESTED** [review]

File: contrib/postgres_proxy/filters/network/source/postgres_decoder.cc (hunk 1 above)
Comment (shiponcs) [reply contrib/postgres_proxy/filters/network/source/postgres_decoder.cc:459]: It is SN or SS depending on whether the upstream server supports TLS or not.

**Comment by wasd171** [comment]
Would love to see it merged! We want to use envoy to accept TLS connections from the downstream, terminate them, and establish new TLS connections to the upstream, similar to the https://github.com/envoyproxy/envoy/issues/27808. Currently this is not supported by the `envoy.filters.network.postgres_proxy`, even though the docs state otherwise

**Comment by cpakulski** [comment]
@shiponcs Thanks for troubleshooting the issue. I debugged the code and while this PR fixes the problem the root cause was somewhere else.
I opened another PR which fixes the root cause: #30194 and added integration test to simulate upstream and downstream SSL.

Please close this PR. Thanks!

**Pull request closed without merging**